cilium/CFP-32820-improve-scalability-of-health-probing.md

```diff
@@ -0,0 +1,149 @@
+
+#### HTTP Probe
+
+The HTTP probe runs in a for loop and [probes each IP](https://github.com/cilium/cilium/blob/fac90aa858c86eefd2c4bdccef8b56abf24a7b58/pkg/health/server/prober.go#L349). Again, for consistency in design, we may modify this probe using [time/rate](https://pkg.go.dev/golang.org/x/time/rate) to run the probe in accordance with the configured requests per second.
+
+### Other Details
+
+#### Connectivity Status Cache Update
+The connectivity cache is currently updated on the condition that the current health report start time is before the health report being cached, at the end of the ProbeInterval period.
+
+With the ProbeInterval set to longer than 60s, it may make more sense to update the connectivity cache more frequently than the ProbeInterval, so that users see some refreshed metrics based on what has been collected at a given point.
+
+A possible solution for this may be to continue refreshing the connectivity cache at intervals of 60 seconds. Instead of storing the start time for the health report as a whole, we may store the health report start time on a per-node basis. Then, every 60 seconds, we make a call to refresh the cache, and update metrics for nodes iff the start time for the node’s last stored metric is before the start time of the current health report.
+
+#### Metrics Update Period
+
+As a consequence of this proposal, health probe updates may take longer and be more unpredictably-timed for users. In order to keep users aware of how long it will take to receive accurate metrics for the known nodes in the cluster, we can update the output of `cilium-health status` to also report the current probe interval and/or the most recent timestamp at which metrics for all known nodes were updated.
+
+#### Probe Timeouts and Errors
+
+The HTTP probe is currently set to have a timeout of 30 seconds. The ICMP probe has MaxRTT set to 60 seconds, which acts as the timeout for the pings. Both timeouts seem larger than necessary. The HTTP probe timeout can be re-configured to 10 seconds, which is a reasonable timeout and accounts for the sizes of the maximum bucket configured for the histogram metric `node_health_connectivity_latency_seconds`.
+
+However, for the ICMP probe, by nature of the library `servak/go-fastping`, the probe interval value and max RTT cannot easily be separated. This is one reason to consider changing to the library `prometheus-community/pro-bing`, which will allow us to easily configure different Timeout and Interval values, allowing us to set the timeout to 10 seconds and the interval to the probe interval value.
+
+#### Synchronous Probe Behavior
+Currently, a synchronous probe is run in two situations:
+1. When the health server starts running
+2. When an HTTP GET request is made to the health server for /status/probe, usually by a call to `cilium-health status --probe`
```

Comment:
This makes sense to me, thanks for sharing your perspective! I agree that it feels like it may defeat the purpose of this CFP a bit. :) I will adjust the document to intend to deprecate this synchronous probe for now.